libopae++/include/opaec++/mmio.h

```diff
@@ -0,0 +1,56 @@
+// Copyright(c) 2017, Intel Corporation
+//
+// Redistribution  and  use  in source  and  binary  forms,  with  or  without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of  source code  must retain the  above copyright notice,
+//   this list of conditions and the following disclaimer.
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+// * Neither the name  of Intel Corporation  nor the names of its contributors
+//   may be used to  endorse or promote  products derived  from this  software
+//   without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE
+// IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
+// LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
+// CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
+// SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+// INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+// CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+#pragma once
+#include <memory>
+
+namespace opae
+{
+namespace fpga
+{
+namespace io
+{
+
+class mmio
+{
+public:
+    typedef std::shared_ptr<mmio> ptr_t;
+
+    virtual ~mmio() {}
+
+    virtual bool write_mmio32(uint32_t offset, uint32_t value) = 0;
+
+    virtual bool write_mmio64(uint32_t offset, uint64_t value) = 0;
+
+    virtual bool read_mmio32(uint32_t offset, uint32_t & value) const = 0;
+
+    virtual bool read_mmio64(uint32_t offset, uint64_t & value) const = 0;
+
+    virtual uint8_t * mmio_pointer(uint32_t offset) const = 0;
```

Comment:
Should this be volatile?

Reply:
Interesting point.  That would probably save a cast to volatile on the receiving end. Will make the change.

Reply:
Won't the underlying pointer be volatile? Isn't that enough?

Reply:
I think that client code might typically do this:

volatile struct my_afu_mmio *p = (volatile struct my_afu_mmio *) region.mmio_pointer()

or else use the base+offset pattern directly. In the latter case, I think it would save a typecast.